File: packages/sentinel-py/src/ameva_sentinel/adapters/enforcement.py
"""
Explicit consumer-controlled enforcement middleware for Python ASGI/WSGI.
Strictly requires consumer decision callback; never applies default HTTP blocking or status codes.
"""

import json
from typing import Callable, Optional, Any, Dict
from ..core.evaluator import SentinelCostGuardEvaluator
from ..core.budget_types import RequestCostContext, VerifiedPrincipal
from ..core.guards import RequestShapeGuard
from ..trusted_proxy import TrustedProxyPolicy, inspect_client_address


class SentinelASGIEnforcer:
    def __init__(
        self,
        app: Any,
        decide: Callable[[dict], dict],
        evaluator: Optional[SentinelCostGuardEvaluator] = None,
        principal_resolver: Optional[Callable[[dict], Optional[VerifiedPrincipal]]] = None,
        trusted_proxy_policy: Optional[TrustedProxyPolicy] = None,
    ):
        if not callable(decide):
            raise TypeError("SentinelASGIEnforcer requires a mandatory callable `decide`.")
        self.app = app
        self.decide = decide
        self.evaluator = evaluator or SentinelCostGuardEvaluator()
        self.principal_resolver = principal_resolver
        self.trusted_proxy_policy = trusted_proxy_policy

    async def __call__(self, scope: dict, receive: Callable, send: Callable) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        method = scope.get("method", "GET")
        raw_path_bytes = scope.get("raw_path")
        path = raw_path_bytes.decode("latin1").split("?")[0] if raw_path_bytes else scope.get("path", "/")

        request_inspection = RequestShapeGuard.inspect_path(path)

        headers_dict = {}
        for k, v in scope.get("headers", []):
            try:
                headers_dict[k.decode("latin1")] = v.decode("latin1")
            except Exception:
                pass

        client_info = scope.get("client")
        socket_ip = client_info[0] if client_info and len(client_info) > 0 else None
        client_address_inspection = inspect_client_address(
            socket_remote_address=socket_ip,
            headers=headers_dict,
            policy=self.trusted_proxy_policy,
        )

        query_string = scope.get("query_string", b"").decode("utf-8")
        from urllib.parse import parse_qs
        page_size, series_count, time_buckets = None, None, None
        if query_string:
            qs = parse_qs(query_string)
            if "pageSize" in qs:
                try:
                    val = int(qs["pageSize"][0])
                    if val > 0:
                        page_size = val
                except ValueError:
                    pass
            if "seriesCount" in qs:
                try:
                    val = int(qs["seriesCount"][0])
                    if val > 0:
                        series_count = val
                except ValueError:
                    pass
            if "timeBuckets" in qs:
                try:
                    val = int(qs["timeBuckets"][0])
                    if val > 0:
                        time_buckets = val
                except ValueError:
                    pass

        principal = None
        if self.principal_resolver:
            principal = self.principal_resolver(scope)
        elif "sentinel.principal" in scope and isinstance(scope["sentinel.principal"], VerifiedPrincipal):
            principal = scope["sentinel.principal"]

        ctx = RequestCostContext(
            method=method,
            path=path,
            page_size=page_size,
            series_count=series_count,
            time_buckets=time_buckets,
            principal=principal,
            network_key=client_address_inspection.client_address or client_address_inspection.socket_address or "unknown",
        )

        decision = None
        if request_inspection.accepted_by_inspector:
            decision = await self.evaluator.evaluate_async(ctx)

        sentinel_context = {
            "request_inspection": request_inspection,
            "client_address_inspection": client_address_inspection,
            "decision": decision,
            "request_context": ctx,
        }

        scope["ameva_sentinel"] = sentinel_context

        outcome = self.decide(sentinel_context)
        if hasattr(outcome, "__await__"):
            outcome = await outcome

        if outcome and outcome.get("action") == "respond":
            status = outcome.get("status", 400)
            headers = []
            if "headers" in outcome:
                for k, v in outcome["headers"].items():
                    headers.append((k.encode("utf-8"), str(v).encode("utf-8")))

            payload = b""
            if "body" in outcome:
                body = outcome["body"]
                if isinstance(body, (dict, list)):
                    payload = json.dumps(body).encode("utf-8")
                    headers.append((b"content-type", b"application/json"))
                elif isinstance(body, str):
                    payload = body.encode("utf-8")
                elif isinstance(body, bytes):
                    payload = body

            headers.append((b"content-length", str(len(payload)).encode("utf-8")))

            await send({"type": "http.response.start", "status": status, "headers": headers})
            await send({"type": "http.response.body", "body": payload})
            return

        await self.app(scope, receive, send)


class SentinelWSGIEnforcer:
    def __init__(
        self,
        app: Any,
        decide: Callable[[dict], dict],
        evaluator: Optional[SentinelCostGuardEvaluator] = None,
        principal_resolver: Optional[Callable[[dict], Optional[VerifiedPrincipal]]] = None,
        trusted_proxy_policy: Optional[TrustedProxyPolicy] = None,
    ):
        if not callable(decide):
            raise TypeError("SentinelWSGIEnforcer requires a mandatory callable `decide`.")
        self.app = app
        self.decide = decide
        self.evaluator = evaluator or SentinelCostGuardEvaluator()
        self.principal_resolver = principal_resolver
        self.trusted_proxy_policy = trusted_proxy_policy

    def __call__(self, environ: dict, start_response: Callable) -> Any:
        method = environ.get("REQUEST_METHOD", "GET")
        raw_uri = environ.get("RAW_URI") or environ.get("REQUEST_URI") or environ.get("PATH_INFO", "/")
        path = raw_uri.split("?")[0]

        request_inspection = RequestShapeGuard.inspect_path(path)

        headers = {}
        for key, value in environ.items():
            if key.startswith("HTTP_"):
                header_name = key[5:].replace("_", "-").lower()
                headers[header_name] = value

        client_address_inspection = inspect_client_address(
            socket_remote_address=environ.get("REMOTE_ADDR"),
            headers=headers,
            policy=self.trusted_proxy_policy,
        )

        query_string = environ.get("QUERY_STRING", "")
        from urllib.parse import parse_qs
        page_size, series_count, time_buckets = None, None, None
        if query_string:
            qs = parse_qs(query_string)
            if "pageSize" in qs:
                try:
                    val = int(qs["pageSize"][0])
                    if val > 0:
                        page_size = val
                except ValueError:
                    pass
            if "seriesCount" in qs:
                try:
                    val = int(qs["seriesCount"][0])
                    if val > 0:
                        series_count = val
                except ValueError:
                    pass
            if "timeBuckets" in qs:
                try:
                    val = int(qs["timeBuckets"][0])
                    if val > 0:
                        time_buckets = val
                except ValueError:
                    pass

        principal = None
        if self.principal_resolver:
            principal = self.principal_resolver(environ)
        elif "sentinel.principal" in environ and isinstance(environ["sentinel.principal"], VerifiedPrincipal):
            principal = environ["sentinel.principal"]

        ctx = RequestCostContext(
            method=method,
            path=path,
            page_size=page_size,
            series_count=series_count,
            time_buckets=time_buckets,
            principal=principal,
            network_key=client_address_inspection.client_address or client_address_inspection.socket_address or "unknown",
        )

        decision = None
        if request_inspection.accepted_by_inspector:
            decision = self.evaluator.evaluate_sync(ctx)

        sentinel_context = {
            "request_inspection": request_inspection,
            "client_address_inspection": client_address_inspection,
            "decision": decision,
            "request_context": ctx,
        }

        environ["ameva_sentinel"] = sentinel_context

        outcome = self.decide(sentinel_context)

        if outcome and outcome.get("action") == "respond":
            status = outcome.get("status", 400)
            status_text = f"{status} Response"
            headers_list = []
            if "headers" in outcome:
                for k, v in outcome["headers"].items():
                    headers_list.append((k, str(v)))

            payload = b""
            if "body" in outcome:
                body = outcome["body"]
                if isinstance(body, (dict, list)):
                    payload = json.dumps(body).encode("utf-8")
                    headers_list.append(("Content-Type", "application/json"))
                elif isinstance(body, str):
                    payload = body.encode("utf-8")
                elif isinstance(body, bytes):
                    payload = body

            headers_list.append(("Content-Length", str(len(payload))))
            start_response(status_text, headers_list)
            return [payload]

        return self.app(environ, start_response)

File: packages/sentinel/src/adapters/enforcement.ts
/**
 * @file enforcement.ts
 * Explicit consumer-controlled enforcement layer for AMEVA-Sentinel.
 * Strictly requires consumer decision callback; never applies default HTTP blocking or status codes.
 */

import {
  SentinelCostGuardEvaluator,
  VerifiedPrincipal,
  RequestCostContext,
  RequestShapeGuard,
  TrustedProxyPolicy,
  inspectClientAddress,
  CostGuardDecision,
  RequestInspection,
  ClientAddressInspection
} from '@ameva/sentinel-risk-core';

export interface SentinelContext {
  requestInspection: RequestInspection;
  clientAddressInspection: ClientAddressInspection;
  decision?: CostGuardDecision;
  requestContext: RequestCostContext;
}

export interface EnforcementDecision {
  action: 'continue' | 'respond';
  status?: number;
  headers?: Record<string, string>;
  body?: unknown;
}

export type EnforcementPolicy = (
  context: SentinelContext
) => EnforcementDecision | Promise<EnforcementDecision>;

export interface ExpressEnforcerOptions {
  decide: EnforcementPolicy;
  evaluator?: SentinelCostGuardEvaluator;
  principalResolver?: (req: any) => VerifiedPrincipal | undefined;
  trustedProxyPolicy?: Partial<TrustedProxyPolicy>;
}

export function createExpressSentinelEnforcer(options: ExpressEnforcerOptions) {
  if (!options || typeof options.decide !== 'function') {
    throw new TypeError('createExpressSentinelEnforcer requires a mandatory options.decide callback.');
  }

  const evaluator = options.evaluator || new SentinelCostGuardEvaluator();
  const principalResolver = options.principalResolver;
  const trustedProxyPolicy = options.trustedProxyPolicy;
  const decide = options.decide;

  return async function expressSentinelEnforcer(req: any, res: any, next: any) {
    const rawPath = req.originalUrl ? req.originalUrl.split('?')[0] : (req.path || req.url || '/');
    const requestInspection = RequestShapeGuard.inspectPath(rawPath);

    const clientAddressInspection = inspectClientAddress({
      socketRemoteAddress: req.socket?.remoteAddress || req.connection?.remoteAddress || req.ip,
      headers: req.headers
    }, trustedProxyPolicy);

    let pageSize: number | undefined;
    let seriesCount: number | undefined;
    let timeBuckets: number | undefined;

    if (req.query?.pageSize) {
      const val = parseInt(String(req.query.pageSize), 10);
      if (!isNaN(val) && val > 0) pageSize = val;
    }
    if (req.query?.seriesCount) {
      const val = parseInt(String(req.query.seriesCount), 10);
      if (!isNaN(val) && val > 0) seriesCount = val;
    }
    if (req.query?.timeBuckets) {
      const val = parseInt(String(req.query.timeBuckets), 10);
      if (!isNaN(val) && val > 0) timeBuckets = val;
    }

    let principal: VerifiedPrincipal | undefined;
    if (principalResolver) {
      principal = principalResolver(req);
    } else if (req['sentinel.principal']?.authenticated) {
      principal = req['sentinel.principal'];
    }

    const requestContext: RequestCostContext = {
      method: req.method || 'GET',
      path: rawPath,
      pageSize,
      seriesCount,
      timeBuckets,
      principal,
      sessionId: req.sessionID || req.cookies?.['session_id'],
      asn: req.asn ? Number(req.asn) : undefined,
      tenantId: req.headers?.['x-tenant-id'],
      networkKey: clientAddressInspection.clientAddress || clientAddressInspection.socketAddress || 'unknown'
    };

    let decision: CostGuardDecision | undefined;
    if (requestInspection.acceptedByInspector) {
      decision = await evaluator.evaluate(requestContext);
    }

    const sentinelContext: SentinelContext = {
      requestInspection,
      clientAddressInspection,
      decision,
      requestContext
    };

    req.sentinel = sentinelContext;

    const outcome = await decide(sentinelContext);

    if (outcome && outcome.action === 'respond') {
      const status = outcome.status || 400;
      if (outcome.headers) {
        for (const [k, v] of Object.entries(outcome.headers)) {
          res.setHeader(k, v);
        }
      }
      if (outcome.body !== undefined) {
        return res.status(status).json(outcome.body);
      }
      return res.status(status).end();
    }

    return next();
  };
}

export interface FastifyEnforcerOptions {
  decide: EnforcementPolicy;
  evaluator?: SentinelCostGuardEvaluator;
  principalResolver?: (req: any) => VerifiedPrincipal | undefined;
  trustedProxyPolicy?: Partial<TrustedProxyPolicy>;
}

export function createFastifySentinelEnforcer(options: FastifyEnforcerOptions) {
  if (!options || typeof options.decide !== 'function') {
    throw new TypeError('createFastifySentinelEnforcer requires a mandatory options.decide callback.');
  }

  const evaluator = options.evaluator || new SentinelCostGuardEvaluator();
  const principalResolver = options.principalResolver;
  const trustedProxyPolicy = options.trustedProxyPolicy;
  const decide = options.decide;

  return async function fastifySentinelEnforcer(req: any, reply: any, done?: any) {
    const rawPath = req.raw?.url ? req.raw.url.split('?')[0] : (req.url || '/');
    const requestInspection = RequestShapeGuard.inspectPath(rawPath);

    const clientAddressInspection = inspectClientAddress({
      socketRemoteAddress: req.raw?.socket?.remoteAddress || req.ip,
      headers: req.headers
    }, trustedProxyPolicy);

    let pageSize: number | undefined;
    let seriesCount: number | undefined;
    let timeBuckets: number | undefined;

    const query = req.query || {};
    if (query.pageSize) {
      const val = parseInt(String(query.pageSize), 10);
      if (!isNaN(val) && val > 0) pageSize = val;
    }
    if (query.seriesCount) {
      const val = parseInt(String(query.seriesCount), 10);
      if (!isNaN(val) && val > 0) seriesCount = val;
    }
    if (query.timeBuckets) {
      const val = parseInt(String(query.timeBuckets), 10);
      if (!isNaN(val) && val > 0) timeBuckets = val;
    }

    let principal: VerifiedPrincipal | undefined;
    if (principalResolver) {
      principal = principalResolver(req);
    } else if (req['sentinel.principal']?.authenticated) {
      principal = req['sentinel.principal'];
    }

    const requestContext: RequestCostContext = {
      method: req.method || 'GET',
      path: rawPath,
      pageSize,
      seriesCount,
      timeBuckets,
      principal,
      sessionId: req.cookies?.['session_id'],
      tenantId: req.headers?.['x-tenant-id'],
      networkKey: clientAddressInspection.clientAddress || clientAddressInspection.socketAddress || 'unknown'
    };

    let decision: CostGuardDecision | undefined;
    if (requestInspection.acceptedByInspector) {
      decision = await evaluator.evaluate(requestContext);
    }

    const sentinelContext: SentinelContext = {
      requestInspection,
      clientAddressInspection,
      decision,
      requestContext
    };

    req.sentinel = sentinelContext;

    const outcome = await decide(sentinelContext);

    if (outcome && outcome.action === 'respond') {
      const status = outcome.status || 400;
      if (outcome.headers) {
        for (const [k, v] of Object.entries(outcome.headers)) {
          reply.header(k, v);
        }
      }
      if (outcome.body !== undefined) {
        return reply.code(status).send(outcome.body);
      }
      return reply.code(status).send();
    }

    if (typeof done === 'function') {
      done();
    }
  };
}
